### PR TITLE
fix the bug of reward weight setting if checkpoint loaded

### PR DESCRIPTION
**Pull request by hnyu**
Turns out that for playing from a checkpoint, the reward weight is not correctly set, because checkpoint loading happens after class init.

### CODE REVIEW

File: alf/algorithms/lagrangian_reward_weight_algorithm.py [hunk 1]
```diff
@@ -103,7 +102,7 @@ def _inv_softplus(self, tensor):
     def reward_weights(self):
         """Return the detached reward weights. These weights are expected not to
         be changed by external code."""
-        return self._reward_weights.detach().clone()
+        return F.softplus(self._lambdas).detach().clone()
```
Comment (hnyu) [inline alf/algorithms/lagrangian_reward_weight_algorithm.py:105]: This function can be called after class init so if self._lambdas is updated by checkpoint, we should always return the newest value. For simplicity, we don't want a variable to cache its latest value.

File: alf/algorithms/agent.py [hunk 2]
```diff
@@ -393,6 +395,12 @@ def after_train_iter(self, experience, train_info: AgentInfo = None):
         algorithms = list(filter(lambda a: a is not None, algorithms))
         self._agent_helper.after_train_iter(algorithms, experience, train_info)
 
+        # Strictly speaking, if we load a checkpoint for training, then in the
+        # first iteration ``self._rl_algorithm`` might have the wrong reward
+        # weights because the function call below is called only after the first
+        # iteration. But for simplicity (there is no easy way to set the weights
+        # after loading a checkpoint while before the first iteration) here we
+        # ignore its effect.
```
Comment (emailweixu) [inline alf/algorithms/agent.py]: All the changes to parameters will be overridden by checkpoint.  See Trainer._restore_checkpoint(). But there will be some experiences written to replay buffer (if replay buffer is not checkpointed).

Comment (hnyu) [reply alf/algorithms/agent.py]: > All the changes to parameters will be overridden by checkpoint. See Trainer._restore_checkpoint(). But there will be some experiences written to replay buffer (if replay buffer is not checkpointed).

After loading a checkpoint, the first thing is that self._reward_weight_algorithms.reward_weights is updated because it's from a nn.Parameter. Then we need to call set_reward_weights on self._rl_algorithm (otherwise it's all ones by default and it is not nn.Parameter in RLAlgorithm).  Even though there are some experiences written to replay buffer in the first iteration (generated by wrong reward weights), I think it's generally OK for an off-policy algorithm.

File: alf/algorithms/agent.py [hunk 3]
```diff
@@ -175,9 +175,6 @@ def __init__(self,
             reward_weight_algorithm = reward_weight_algorithm_cls(
                 reward_spec=reward_spec, debug_summaries=debug_summaries)
             agent_helper.register_algorithm(reward_weight_algorithm, "rw")
-            # Initialize the reward weights of the rl algorithm
-            rl_algorithm.set_reward_weights(
-                reward_weight_algorithm.reward_weights)
```
Comment (emailweixu) [inline alf/algorithms/agent.py]: why do you remove these lines?

Comment (hnyu) [reply alf/algorithms/agent.py]: > why do you remove these lines?

reverted

**Review by emailweixu — APPROVED** [review]
